### PR TITLE
Follow Debian third party repository best practices 

### PR DESCRIPTION
**Pull request by jooola**
Try to follow the Debian third repository best practices.

This adds an optional `mariadb_use_official_repo_deb822` to use the deb822 format for the sources file. 

See:
- https://wiki.debian.org/DebianRepository/UseThirdParty
- https://repolib.readthedocs.io/en/latest/deb822-format.html

### CODE REVIEW

File: tasks/setup_debian.yml [hunk 1]
```diff
@@ -10,26 +10,63 @@
...
+        owner: root
+        group: root
+        mode: "0755"
+
+    - name: Fetch mariadb repository key
       ansible.builtin.get_url:
-        url: "https://mariadb.org/mariadb_release_signing_key.asc"
-        dest: "/etc/apt/trusted.gpg.d/mariadb.asc"
+        url: https://mariadb.org/mariadb_release_signing_key.asc
+        dest: /etc/apt/keyrings/mariadb-archive-keyring.asc
```
Comment (jooola) [inline tasks/setup_debian.yml:24]: Should the role be responsible for cleaning up the previous key ? Or should this be documented in the changelogs ?

File: tasks/setup_debian.yml [hunk 2]
```diff
@@ -10,26 +10,63 @@
...
         group: root
         mode: "0644"
+      register: fetch_repository_key
+
+    - name: Dearmor mariadb repository key
+      when: fetch_repository_key is changed
+      ansible.builtin.command: >
+        gpg /etc/apt/keyrings/mariadb-archive-keyring.asc
+        --dearmor
+        --output /etc/apt/keyrings/mariadb-archive-keyring.gpg
```
Comment (jooola) [inline tasks/setup_debian.yml:37]: This command and the one above could be simplified to:

```yml
- ansible.builtin.shell: >
    curl -sSL https://mariadb.org/mariadb_release_signing_key.asc
    | gpg --dearmor --output /etc/apt/keyrings/mariadb-archive-keyring.gpg
```
But:
- remove the temporary file at /etc/apt/keyrings/mariadb-archive-keyring.asc 
- raises some linting errors
- can't easily check if the key changed or not

File: tasks/setup_debian.yml [hunk 3]
```diff
@@ -10,26 +10,63 @@
...
+          {{ distribution }} \
+          main
+        owner: root
+        group: root
+        mode: "0644"
+
+    - name: Setup mariadb repository sources entry
+      when: mariadb_use_official_repo_deb822
+      ansible.builtin.copy:
+        dest: /etc/apt/sources.list.d/mariadb.sources
```
Comment (jooola) [inline tasks/setup_debian.yml:61]: Should the role be responsible for cleaning up the files when changing `mariadb_use_official_repo_deb822` ?

**Review by jooola — COMMENTED** [review]
I'd rather put the deb822 feature behind a feature flag for now.

A future PR change could add more guessing for the `mariadb_use_official_repo_deb822` variable, using distro specific variables.

File: tasks/setup_debian.yml (hunk 3 above)
Comment (fauust) [reply tasks/setup_debian.yml:59]: Nope, the role should consider a from scratch deployment (cleaning scenarios are too complex IMO).

**Comment by fauust** [comment]
Any chance this PR could also fix the [sid problem](https://github.com/fauust/ansible-role-mariadb/actions/runs/3703846003/jobs/6275776380)?

I don't know if it's on purpose but debian sid images have `VERSION_CODENAME=bookworm`, see bellow:
```console
❯ podman run -it debian:sid-slim bash -c "cat /etc/os-release"
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

And ansible probably use it to guess `{{ ansible_distribution }}`.

**Comment by jooola** [comment]
> Any chance this PR could also fix the [sid problem](https://github.com/fauust/ansible-role-mariadb/actions/runs/3703846003/jobs/6275776380)?

This is solved in a27a98a1eb9da0095c940e27dae89de2bb63736a

Since I could squeeze some extra changes in this PR, I went all the way and fixed some extra things that bothered my while working on this.

This is ready for review, I don't whether the check mode is still working as expected though. Maybe a dumb CI job for this is worth it ?

**Comment by jooola** [comment]
Also I don't how to properly test the new feature that I have introduced, I don't want to add more CI jobs, so maybe we target some specific jobs that will use this particular source format ? Say all the jobs that installs mariadb 10.6 ?

File: defaults/main.yml [hunk 4]
```diff
@@ -7,6 +7,8 @@ ansible_python_interpreter: /usr/bin/python3
 mariadb_use_official_repo: false
 mariadb_use_official_repo_url: https://deb.mariadb.org
 mariadb_use_official_repo_version: 10.6
+# see: https://repolib.readthedocs.io/en/latest/deb822-format.html
```
Comment (fauust) [inline defaults/main.yml:10]: https://manpages.debian.org/stable/apt/sources.list.5.en.html#DEB822-STYLE_FORMAT is official doc.

File: README.md [hunk 5]
```diff
@@ -92,6 +92,7 @@ Available variables are listed below, along with default values (see
 mariadb_use_official_repo: false
 mariadb_use_official_repo_url: https://deb.mariadb.org
 mariadb_use_official_repo_version: "10.10"
+mariadb_use_official_repo_deb822: false
```
Comment (fauust) [inline README.md:95]: Can you use something more clear (like `mariadb_source_list_format_deb822`) and this is not related to using MDBF official vs debian repositories.

File: tasks/setup_debian.yml [hunk 6]
```diff
@@ -10,28 +10,62 @@
...
+      ansible.builtin.file:
+        path: /etc/apt/keyrings
+        state: directory
+        owner: root
+        group: root
+        mode: "0755"
+
+    - name: Fetch mariadb repository key
+      # change is always triggered in check mode
+      when: not ansible_check_mode
```
Comment (fauust) [inline tasks/setup_debian.yml:23]: For consistency, `when` needs to be at the end (see also other occurrence).

File: tasks/setup_debian.yml [hunk 7]
```diff
@@ -10,28 +10,62 @@
...
+      ansible.builtin.copy:
+        dest: /etc/apt/sources.list.d/mariadb.sources
+        content: |
+          # Ansible managed
+          X-Repolib-Name: MariaDB
+          Types: deb
+          URIs: {{ mariadb_use_official_repo_url }}/{{ mariadb_use_official_repo_version }}/{{ ansible_distribution | lower() }}
+          Suites: {{ repo_distribution_release }}
+          Components: main
+          Signed-By: /etc/apt/keyrings/mariadb-archive-keyring.gpg
```
Comment (fauust) [inline tasks/setup_debian.yml:68]: Can you change for `mariadb-keyring.gpg`, small is beautiful :)

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -12,11 +12,13 @@ repos:
       - id: trailing-whitespace
         name: Check trailing whitespace
         args: [--markdown-linebreak-ext=md]
+
```
Comment (fauust) [inline .pre-commit-config.yaml:15]: Personal taste, correct also bellow..

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -55,15 +59,15 @@ repos:
...
+            css|less|scss
+            graphql|gql|
+            html|
+            js|jsx|ts|tsx|
+            json|
+            md|markdown|mdown|mkdn|
+            mdx|
+            vue|
+            yaml|yml
+          )$
```
Comment (fauust) [inline .pre-commit-config.yaml:73]: Yep, that's cleaner :+1:

**Review by fauust — COMMENTED** [review]
I am wondering if we should add some testing for the new sources.list format in the CI.

**Comment by fauust** [comment]
> Also I don't how to properly test the new feature that I have introduced, I don't want to add more CI jobs, so maybe we target some specific jobs that will use this particular source format ? Say all debian based (apt based) jobs that installs mariadb 10.6 ?

Yep good Q. let me think about. We could add it to "recent" Debian/Ubuntu versions (say for instance Sid and Ubuntu > 22.04)...?

File: README.md (hunk 5 above)
Comment (jooola) [reply README.md:95]: We deploy a repository source only when using the 'use_official_repo', so it seem appropriate to keep the same prefixes.

Maybe ?
```
mariadb_use_official_repo_format_deb822
```

File: tasks/setup_debian.yml (hunk 7 above)
Comment (jooola) [reply tasks/setup_debian.yml:68]: I am following the debian keyrings names, but sure.

image

File: .pre-commit-config.yaml (hunk 8 above)
Comment (jooola) [reply .pre-commit-config.yaml:15]: I would argue it improves readability, but ok.

File: README.md (hunk 5 above)
Comment (fauust) [reply README.md]: yep true, ok

File: tasks/setup_debian.yml (hunk 7 above)
Comment (jooola) [reply tasks/setup_debian.yml]: According to https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_certificate_distribution we diverged a bit with regards to the naming, unless we want to be maniac about this, I would say we are fine with the shorter name.
